### PR TITLE
Custom Instrumentation

### PR DESCRIPTION
**Pull request by bwaresiak**
Task/Issue URL: https://app.asana.com/0/856498667320406/1122248096879715
Tech Design URL:
CC:

**Description**:
Added Custom Instruments target to project. Please check README.md for instruction how to run it. Note, that this is initial implementation and functionality is limited. :)

Features:
- Info about URLs being loaded in the tabs (only successful ones are logged for now).
- Info about performance-related events like Content Blocker updates, DisconnectMe Parsing/loading, data cleaning, tab initialization.
- Info about which requests are blocked and which are not.

**Steps to test this PR**:
1. Ensure README.md looks ok.
2. Ensure that debug JS messaging is only enabled in debug builds.
3. Ensure no uninvited logs are being logged to console in release build.
4. Test and play with it in instruments.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo.xcodeproj/xcshareddata/xcschemes/Instruments.xcscheme [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      selectedLauncherIdentifier = "Xcode.IDEFoundation.Launcher.PosixSpawn"
+      launchStyle = "0"
+      useCustomWorkingDirectory = "NO"
+      ignoresPersistentStateOnLaunch = "NO"
+      debugDocumentVersioning = "YES"
+      debugServiceExtension = "internal"
+      allowLocationSimulation = "YES">
+      <PathRunnable
+         runnableDebuggingMode = "0"
+         FilePath = "/Users/bartek/Downloads/Xcode.app/Contents/Applications/Instruments.app">
```
Comment (brindy) [inline DuckDuckGo.xcodeproj/xcshareddata/xcschemes/Instruments.xcscheme]: User specific path - does this matter?

File: Core/WKWebViewConfigurationExtension.swift [hunk 2]
```diff
@@ -101,6 +101,19 @@ private class Loader {
         } else {
             load(scripts: [ .messaging, .apbfilterES2015 ], forMainFrameOnly: false)
         }
+        
+        if isDebugBuild {
+            let signpostMessaging = "try { webkit.messageHandlers.signpostMessage.postMessage(data); } catch(error) {}"
```
Comment (brindy) [inline Core/WKWebViewConfigurationExtension.swift]: I'm not keen on having JS in strings in our swift code. How about inserting an alternative JS depending on the isDebugBuild flag instead?

**Review by brindy — COMMENTED** [review]
This is very cool! Two comments for you as discussed.

File: DuckDuckGo.xcodeproj/xcshareddata/xcschemes/Instruments.xcscheme (hunk 1 above)
Comment (bwaresiak) [reply DuckDuckGo.xcodeproj/xcshareddata/xcschemes/Instruments.xcscheme]: I'd say not really. I've kept my Xcode in Downloads, after moving it to Applications it has changed the path to more reasonable one. :)

Comment (bwaresiak) [reply DuckDuckGo.xcodeproj/xcshareddata/xcschemes/Instruments.xcscheme]: I've tweaked that a little by putting Xcode.app into /Applications ;) Path seems to make more sense now.

File: Core/WKWebViewConfigurationExtension.swift (hunk 2 above)
Comment (bwaresiak) [reply Core/WKWebViewConfigurationExtension.swift]: Done, along with 'log' message handler.

File: README.md [hunk 3]
```diff
@@ -24,6 +24,20 @@ You can also run the unit tests to do the above and ensure everything seems in o
...
+
+### Instruments
+
+We have Custom Instruments tool to help visualize and track events that happen during runtime.
+
+In order to run it:
+1. Build a Debug version and install it on Simulator/Device.
+2. Select Instruments target and run it on a Mac. New instance of Instruments app will be run that has a grayed out icon indicating that it works in debug mode with custom instruments attached.
+3. Select 'DDG Trace' template or setup a custom one by importing 'DDG Timeline' instrument from Library .
+5. Start recording.
```
Comment (brindy) [inline README.md]: where is number 4? :)

I had to select the simulator and ddg app before starting recording, is that what number 4 is?

**Review by brindy — APPROVED** [review]
One comment in the documentation, but otherwise feel free to go ahead and merge.

File: README.md (hunk 3 above)
Comment (bwaresiak) [reply README.md]: Sounds like a good number 4 to me! 😄

**Comment by bwaresiak** [comment]
Thanks :)

**Pull request merged**